File: js/level/levelgen.js

var FREQUENCY = 1;
var OCTAVES = 16;

LevelGen = Class.extend({
	w: 0,
	h: 0,
	seed: null,

	// Simple smoothing method, gets rid of unwanted blurps
	smooth: function() {
		for (var y = 0; y < this.h; y++) {
			for (var x = 0; x < this.w; x++) {
				var average = 0;
				var times = 0;

				if (x - 1 >= 0) {
					average += this.map[y][x - 1];
					times++;
				}
				if (x + 1 < this.w - 1) {
					average += this.map[y][x + 1];
					times++;
				}
				if (y - 1 >= 0) {
					average += this.map[y - 1][x];
					times++;
				}
				if (y + 1 < this.h - 1) {
					average += this.map[y + 1][x];
					times++
				}

				if (x - 1 >= 0 && y - 1 >= 0) {
					average += this.map[y - 1][x - 1];
					times++;
				}
				if (x + 1 < this.w - 1 && y - 1 >= 0) {
					average += this.map[y - 1][x + 1];
					times++;
				}
				if (x - 1 >= 0 && y + 1 < this.h - 1) {
					average += this.map[y + 1][x - 1];
					times++;
				}
				if (x + 1 < this.w  - 1&& y + 1 < this.h - 1) {
					average += this.map[y + 1][x + 1];
					times++
				}

				average += this.map[y][x];
				times++;

				average /= times;

				this.map[y][x] = average;
			}
		}
	},

	createIsland: function(w, h) {
		this.map = gPerlinNoiseGen.generateNoise(w, h, FREQUENCY, OCTAVES);

		this.w = w;
		this.h = h;

		var particleMap = [];
		for (var y = 0; y < this.h; y++) {
			var row = [];
			for (var x = 0; x < this.w; x++) {
				row[x] = 0;
			}
			particleMap[y] = row;
		}

		for (var i = 0; i < Math.floor(w * h * 0.85); i++) {
			var x = 15 + Math.floor(Math.random() * (this.w - 33));
			var y = 15 + Math.floor(Math.random() * (this.h - 33));
			// console.log("i="+i+" x="+x+" y="+y);
			for (var j = 0; j < Math.floor(w * h * 0.05); j++) {
				particleMap[y][x] += 7;

				if (particleMap[y][x] >= 255) {
					particleMap[y][x] = 255;
				}

				var currentValue  = particleMap[y][x];
				var choices = [];

				if (x - 1 > 0) {
					if (particleMap[y][x - 1] <= currentValue) {
						choices.push('left');
					}
				}
				if (x + 1 < this.w - 1) {
					if (particleMap[y][x + 1] <= currentValue) {
						choices.push('right');
					}
				}
				if (y - 1 > 0) {
					if (particleMap[y - 1][x] <= currentValue) {
						choices.push('up');
					}
				}
				if (y + 1 < this.w - 1) {
					if (particleMap[y + 1][x] <= currentValue) {
						choices.push('down');
					}
				}

				if (choices.length == 0) {
					break;
				}

				var actualChoice = choices[Math.floor(Math.random() * 4)];
				if (actualChoice == 'left') {
					x -= 1;
				} else if (actualChoice == 'right') {
					x += 1;
				} else if (actualChoice == 'up') {
					y -= 1;
				} else if (actualChoice == 'down') {
					y += 1;
				}
			}
		}

		for (var y = 0; y < this.h; y++) {
			for (var x = 0; x < this.w; x++) {
				// console.log("x="+x+" y="+y+" particleMap[y][x]="+particleMap[y][x]);
				this.map[y][x] *= particleMap[y][x] / 255;
			}
		}

		this.smooth();
		// this.smooth();
		var waterline = this.getWaterLine();
		var rockCount = 0;
		var grassCount = 0;
		var sandCount = 0;
		var waterCount = 0;

		for (var y = 0; y < this.h; y++) {
			for (var x = 0; x < this.w; x++) {
				var tile = Math.floor(this.map[y][x]);

				var mapTile = {
					id: 0
				}

				if (tile <= waterline) {
					mapTile.id = gTileLibrary['water'].id;
					waterCount++;
				} else if (tile > waterline && tile <= waterline + 15) {
					mapTile.id = gTileLibrary['sand'].id;
					sandCount++;
				} else if (tile > waterline + 15 && tile <= waterline + 40) {
					mapTile.id = gTileLibrary['grass'].id;
					grassCount++;
				} else if (tile > waterline + 40 && tile <= waterline + 60) {
					mapTile.id = gTileLibrary['tree'].id;
					grassCount++;
				} else {
					mapTile.id = gTileLibrary['rock'].id;
					rockCount++;
				}

				this.map[y][x] = mapTile;
			}
		}
				
		console.log("map total :\n\t"+rockCount+" rock / "+grassCount+" grass / "+sandCount+" sand / "+waterCount+" water");

		return this.map;
	},

	getWaterLine: function() {
		var values = [];

		for (var y = 0; y < this.h; y++) {
			for (var x = 0; x < this.w; x++) {
				values.push(this.map[y][x]);
			}
		}
		values.sort();
		// console.log(values.length + " values / waterline @ " + ((values.length - 1) * 0.60) + " / rounded @ " + Math.floor((values.length - 1) * 0.60));
		return values[Math.floor((values.length - 1) * 0.40)];
	},

	createSurfaceMap: function(w, h) {
		var attempt = 0;

		do {
			var result = this.createIsland(w, h);

			return result;
		} while (true);
	}
});

File: src/xrpg/game/level/LevelGen.java
package xrpg.game.level;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

import xrpg.game.level.tile.Tile;

public abstract class LevelGen {
	private static final int FREQUENCY = 1;
	private static final int OCTAVES = 16;

	// Simple smoothing method, gets rid of unwanted blurps.
	private static void smooth(int w, int h, double[][] map) {
		for (int y = 0; y < h; y++) {
			for (int x = 0; x < w; x++) {
				double average = 0;
				int times = 0;

				if (x - 1 >= 0) {
					average += map[y][x - 1];
					times++;
				}
				if (x + 1 < w - 1) {
					average += map[y][x + 1];
					times++;
				}
				if (y - 1 >= 0) {
					average += map[y - 1][x];
					times++;
				}
				if (y + 1 < h - 1) {
					average += map[y + 1][x];
					times++;
				}

				if (x - 1 >= 0 && y - 1 >= 0) {
					average += map[y - 1][x - 1];
					times++;
				}
				if (x + 1 < w - 1 && y - 1 >= 0) {
					average += map[y - 1][x + 1];
					times++;
				}
				if (x - 1 >= 0 && y + 1 < h - 1) {
					average += map[y + 1][x - 1];
					times++;
				}
				if (x + 1 < w  - 1&& y + 1 < h - 1) {
					average += map[y + 1][x + 1];
					times++;
				}

				average += map[y][x];
				times++;

				average /= times;

				map[y][x] = average;
			}
		}
	}

	private static char[][] createIsland(Random random, int w, int h) {
		double[][] map = PerlinNoise.generateNoise(random, w, h, FREQUENCY, OCTAVES);

		int[][] particleMap = new int[h][w];

		for (int i = 0; i < (int)(w * h * 0.85); i++) {
			int x = 15 + random.nextInt(w - 33);
			int y = 15 + random.nextInt(h - 33);

			// System.out.println("i="+i+" x="+x+" y="+y);
			for (int j = 0; j < Math.floor(w * h * 0.05); j++) {
				particleMap[y][x] += 7;

				if (particleMap[y][x] >= 255) {
					particleMap[y][x] = 255;
				}

				int currentValue  = particleMap[y][x];
				
				List<Integer> choices = new ArrayList<Integer>();
				if (x - 1 > 0) {
					if (particleMap[y][x - 1] <= currentValue) {
						choices.add(0);
					}
				}
				if (x + 1 < w - 1) {
					if (particleMap[y][x + 1] <= currentValue) {
						choices.add(1);
					}
				}
				if (y - 1 > 0) {
					if (particleMap[y - 1][x] <= currentValue) {
						choices.add(2);
					}
				}
				if (y + 1 < w - 1) {
					if (particleMap[y + 1][x] <= currentValue) {
						choices.add(3);
					}
				}

				if (choices.isEmpty()) {
					break;
				}

				int actualChoice = choices.get(random.nextInt(choices.size()));
				if (actualChoice == 0) {
					x -= 1;
				} else if (actualChoice == 1) {
					x += 1;
				} else if (actualChoice == 2) {
					y -= 1;
				} else if (actualChoice == 3) {
					y += 1;
				}
			}
		}

		for (int y = 0; y < h; y++) {
			for (int x = 0; x < w; x++) {
				map[y][x] *= (double)particleMap[y][x] / 255;
				//System.out.println("y="+y+" x="+x+" map[y][x]="+map[y][x]);
			}
		}

		smooth(w, h, map);

		double waterline = getWaterLine(w, h, map);
		long rockCount = 0;
		long grassCount = 0;
		long sandCount = 0;
		long waterCount = 0;
		long treeCount = 0;

		System.out.println("Waterline is @ " + waterline);
		// Generate ground.
		char[][] tileMap = new char[h][w];
		for (int y = 0; y < h; y++) {
			for (int x = 0; x < w; x++) {
				double tile = Math.floor(map[y][x]);

				char mapTile;

				if (tile <= waterline) {
					mapTile = Tile.ms_water.m_tileId;
					waterCount++;
				} else if (tile > waterline && tile <= waterline + 15) {
					mapTile = Tile.ms_sand.m_tileId;
					sandCount++;
				} else if (tile > waterline + 15 && tile <= waterline + 60) {
					mapTile = Tile.ms_grass.m_tileId;
					grassCount++;
				} else {
					mapTile = Tile.ms_rock.m_tileId;
					rockCount++;
				}

				// Add trees.
				if (mapTile == Tile.ms_grass.m_tileId) {
					if (random.nextInt(10) == 0) {
						mapTile = Tile.ms_tree.m_tileId;
						grassCount--;
						treeCount++;
					}
				}

				// Add cacti.
				if (mapTile == Tile.ms_sand.m_tileId) {
					if (random.nextInt(30) == 0) {
						mapTile = Tile.ms_cactus.m_tileId;
						grassCount--;
						treeCount++;
					}
				}

				tileMap[y][x] = mapTile;
			}
		}

		System.out.println("map total :\n\t"+rockCount+" rock / "+grassCount+" grass / "+sandCount+" sand / "+waterCount+" water / "+treeCount+" tree");

		return tileMap;
	}

	public static List<Town> getTowns(Random random, int w, int h, char[][] map) {
		List<Town> towns = new ArrayList<Town>();

		int xt = (int) Math.floor((w / 10) + (random.nextFloat() * w * 0.8));
		int yt = (int) Math.floor((w / 10) + (random.nextFloat() * h * 0.8));

		while (map[yt][xt] == Tile.ms_water.m_tileId || map[yt][xt] == Tile.ms_rock.m_tileId) {
			int dir = random.nextInt(4);
			if (dir == 0) xt++;
			if (dir == 1) xt--;
			if (dir == 2) yt++;
			if (dir == 3) yt--;
		}

		towns.add(new Town(0, xt, yt, random));

		return towns;
	}

	public static double getWaterLine(int w, int h, double[][] map) {
		double[] values = new double[w * h];

		for (int y = 0; y < h; y++) {
			for (int x = 0; x < w; x++) {
				values[y * h + x] = map[y][x];
			}
		}
		
		Arrays.sort(values);
		
		System.out.println(values.length + " values / waterline @ " + ((values.length - 1) * 0.40) + " / rounded @ " + Math.floor((values.length - 1) * 0.40));
		return values[(int) Math.floor((values.length - 1) * 0.40)];
	}

	public static char[][] createSurfaceMap(Random random, int w, int h) {
		return createIsland(random, w, h);
	}
}
